include/xtensor/xmath.hpp

```diff
@@ -2402,6 +2402,105 @@ XTENSOR_INT_SPECIALIZATION_IMPL(FUNC_NAME, RETURN_VAL, unsigned long long);
 
         return eval(sum(trap, {saxis}));
     }
+
+    /**
+     * @ingroup basic_functions
+     * @brief Returns the one-dimensional piecewise linear interpolant to a function with given discrete data points (xp, fp), evaluated at x.
+     *
+     * @param The x-coordinates at which to evaluate the interpolated values (sorted).
+     * @param The x-coordinates of the data points (sorted).
+     * @param The y-coordinates of the data points, same length as xp.
+     * @param Value to return for x < xp[0].
+     * @param Value to return for x > xp[-1]
+     * @return an one-dimensional xarray, same length as x.
+     */
+    template<class E1, class E2, class E3, typename T>
+    inline auto interp(const E1 &x, const E2 &xp, const E3 &fp, T left, T right)
+    {
+        using size_type = detail::common_size_type_t<E1,E2,E3>;
+        using value_type = typename E3::value_type;
+
+        // basic checks
+        #ifdef XTENSOR_ENABLE_ASSERT
+
+            XTENSOR_ASSERT( xp.dimension() == 1 );
+
+            for (size_type i = 1 ; i < x.size() ; ++i)
+            {
+                XTENSOR_ASSERT( x[i] >= x[i - 1] );
+            }
+
+            for (size_type ip = 1 ; ip < xp.size() ; ++ip)
+            {
+                XTENSOR_ASSERT( xp[ip] >= xp[ip - 1] );
+            }
+
+        #endif
+
+        // allocate output
+        auto f = xtensor<value_type, 1>::from_shape(x.shape());
+
+        // counter in "x"
+        size_type i    = 0;
+        size_type imax = x.size();
+
+        // fill f[i] for x[i] <= xp[0]
+        for (; i < x.size() ; ++i)
+        {
+            if (x[i] > xp[0])
+            {
+                break;
+            }
+            f[i] = static_cast<value_type>(left);
+        }
+
+        // fill f[i] for x[-1] >= xp[-1]
+        for (; imax-- > 0 ;)
```

Comment:
It seems that when `imax` reaches 0, `--` will make it loop back to the maximum value.

We need to be very careful about backward loops on unsigned integers.

Reply:
I think @tdegeus had this logic before in his library so I think it should be fine, correct?

Reply:
@SylvainCorlay @wolfv You're both right. In terms of the loop counter it was in fact right. However continued use below the loop did give a problem if the loop was truncated at `imax = 0` (never used it in this way before). I've corrected this, thanks for catching the edge case!

> Note that beyond the test case, I have checked directly against NumPy's `interp` with some large random distribution.